src/vm/arm64/asmhelpers.asm

```diff
@@ -1386,5 +1391,101 @@ CallHelper2
+;__declspec(naked) void F_CALL_CONV JIT_Stelem_Ref(PtrArray* array, unsigned idx, Object* val)
+    LEAF_ENTRY JIT_Stelem_Ref
+    ; We retain arguments as they were passed and use x0 == array x1 == idx x2 == val
+
+    ; check for null array
+    cbz     x0, ThrowNullReferenceException
+
+    ; idx bounds check
+    ldr     x3,[x0,#ArrayBase__m_NumComponents]
+    cmp     x3, x1
+    bls     ThrowIndexOutOfRangeException
+
+    ; fast path to null assignment (doesn't need any write-barriers)
+    cbz     x2, AssigningNull
+
+    ; Verify the array-type and val-type matches before writing
+    ldr     x12, [x0] ; x12 = array MT
+    ldr     x3, [x2] ; x3 = val->GetMethodTable()
+    ldr     x12, [x12, #MethodTable__m_ElementType] ; array->GetArrayElementTypeHandle()
+    cmp     x3, x12
+    beq     JIT_Stelem_DoWrite
+
+    ; Types didnt match but allow writing into an array of objects
+    ldr     x3, =$g_pObjectClass
+    ldr     x3, [x3]  ; x3 = *g_pObjectClass
+    cmp     x3, x12   ; array type matches with Object*
+    beq     JIT_Stelem_DoWrite
+
+    ; array type and val type do not exactly match. Raise frame and do detailed match
+    b       JIT_Stelem_Ref_NotExactMatch
+
+AssigningNull
+    ; Assigning null doesn't need write barrier
+    add     x0, x0, x1, LSL #3           ; x0 = x0 + (x1 x 8) = array->m_array[idx]
+    str     x2, [x0, #PtrArray__m_Array] ; array->m_array[idx] = val
+    ret
+
+ThrowNullReferenceException
+    ; Tail call JIT_InternalThrow(NullReferenceException)
+    ldr     x0, =CORINFO_NullReferenceException_ASM
+    b       JIT_InternalThrow
+
+ThrowIndexOutOfRangeException
+    ; Tail call JIT_InternalThrow(NullReferenceException)
+    ldr     x0, =CORINFO_IndexOutOfRangeException_ASM
+    b       JIT_InternalThrow
+
+   LEAF_END 
+
+; ------------------------------------------------------------------
+; __declspec(naked) void F_CALL_CONV JIT_Stelem_Ref_NotExactMatch(PtrArray* array,
+;                                                       unsigned idx, Object* val)
+;   x12 = array->GetArrayElementTypeHandle()
+;
+    NESTED_ENTRY JIT_Stelem_Ref_NotExactMatch    
+	PROLOG_SAVE_REG_PAIR           fp, lr, #-0x48!    
+	stp     x0, x1, [sp, #16]
+	str     x2, [sp, #32]
+
+    ; allow in case val can be casted to array element type
+    ; call ObjIsInstanceOfNoGC(val, array->GetArrayElementTypeHandle())
+    mov     x1, x12 ; array->GetArrayElementTypeHandle()
+    mov     x0, x2
+    bl      ObjIsInstanceOfNoGC
+    cmp     x0, TypeHandle_CanCast
+    beq     DoWrite             ; ObjIsInstance returned TypeHandle::CanCast
+
+    ; check via raising frame
+NeedFrame
+    add     x1, sp, #16             ; x1 = &array
+    add     x0, sp, #32             ; x0 = &val
+
+    bl      ArrayStoreCheck ; ArrayStoreCheck(&val, &array)
+
+DoWrite        
+	ldp     x0, x1, [sp], #16
+	ldr     x2, [sp], #32	
+    EPILOG_RESTORE_REG_PAIR           fp, lr, #0x48!
+    b JIT_Stelem_DoWrite
```

Comment:
Should this be `EPILOG_BRANCH` to generate the right unwind info?

Reply:
I agree. Thanks for pointing it. I am correcting it.